Automated testing in WordPress/Codeception.md

```diff
@@ -2,18 +2,307 @@ Codeception collects and shares recommended practices and solutions for testing
+
+#### Functional tests
+
+The functional tests test the __functionality from the perspective of a developer__.
+
+Here you could test some custom validation rules, Ajax requests and similar.
+
+#### Integration tests
+
+These tests will __test code modules in the context of a WordPress app__.
+
+You would test each of your functionality (usually split across different functions) using integration tests. For instance, say we have created a custom rest route that will handle the contact form functionality - you would POST some data to it, and, depending on the data, expect some result. During this test a lot of your code functionalities will be tested - rest setup and existence of correct route, route validation, your business logic that handles data sent to the route, email sending, etc.
+
+In your `tests/wpunit/Routes` you could create a `ContactRouteTest.php` file
+
+```php
+<?php
+
+declare(strict_types=1);
+
+namespace Tests\WPUnit\Routes;
+
+use Codeception\TestCase\WPTestCase;
+use MyPlugin\Routes\Endpoints\Contact;
+
+class ContactRouteTest extends WPTestCase
+{
+  private $request;
+  private $route_name;
+
+  public function setUp()
+  {
+    parent::setUp();
+
+    // Set up a REST server instance.
+    global $wp_rest_server;
+
+    $this->server = $wp_rest_server = new \WP_REST_Server;
+    do_action( 'rest_api_init' );
+
+    // Helper.
+    $this->route_name = '/' . Contact::NAMESPACE_NAME . Contact::VERSION . Contact::ROUTE_NAME;
+
+    $this->request = new \WP_REST_Request( 'POST', $this->route_name );
+  }
+
+  public function tearDown()
+  {
+    // Test cleanup.
+    global $wp_rest_server;
+    $wp_rest_server = null;
+
+    parent::tearDown();
+  }
+
+  public function testRouteIsRegistered()
+  {
+    $routes = $this->server->get_routes();
+
+    $this->assertArrayHasKey($this->route_name, $routes);
+  }
+
+  public function testRequestWasSuccessful()
+  {
+    $this->request->add_header('Content-Type', 'application/json;charset=utf-8');
+    $this->request->set_body('{"name":"Test Testovski", "email":"test.email@email.com", "message":"Sample Message", "recaptcha":"some_random_string"}');
+
+    $response = $this->server->dispatch( $this->request );
+    $data = $response->get_data();
+
+    $this->assertEquals( $data['http_response'], 204 );
+    $this->assertEquals( $data['message'], 'Contact us email is sent. You will be contacted soon.' );
+  }
+
+  public function testHoneypotResponseIsTriggered()
+  {
+    $this->request->add_header('Content-Type', 'application/json;charset=utf-8');
+    $this->request->set_body('{"website":"true", "name":"Test Testovski", "email":"test.email@email.com", "message":"Sample Message", "recaptcha":"some_random_string"}');
+
+    $response = $this->server->dispatch( $this->request );
+    $data = $response->get_data();
+
+    $this->assertEquals( $data['code'], 403 );
+    $this->assertEquals( $data['message'], 'bot_detected' );
```

Comment:
This can be much more abstracted (doesn't depend on `WPTestCase` either) like so:

```php
  /**
   * Ping the route and make sure it returns correct HTTP code and it returns JSON
   *
   * @param \ApiTester $I Codeception ApiTester object.
   * @return void
   */
  public function routeShouldWork(\UnitTester $I)
  {
    $I->sendPOST($this->getRoutePath(), $this->requiredBaseParameters());
    $I->seeResponseCodeIs(200);
    $I->seeResponseIsJson();
    $I->seeResponseMatchesJsonType($this->expectedResponseMatchJson());
  }
```
https://github.com/infinum/wordpress-tarisio/blob/master/tests/integration/Routes/BaseEndpointCase.php.

The snippet above works assuming your `UnitTester` is setup for Route testing using `PhpBrowser` - see https://github.com/infinum/wordpress-tarisio/blob/master/tests/unit.suite.yml.

Alternatively you can setup a specific `ApiTester` for this (this seems to be the preferred method, see https://codeception.com/docs/10-APITesting)

Reply:
I think in the case of wp-browser it would be `WpunitTester`, but I get what you are saying. This test is taken from one of the projects where I have the integration tests set up (native ones), and I followed the official docs: https://wpbrowser.wptestkit.dev/levels-of-testing#integration-tests

I could try to use this abstraction (I'm sure it would work), I'd just like to actually write such test first to be sure 😅

For this example, I know that it works, since it's how rest is tested in basic WP integration tests 🙂